framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -974,6 +976,10 @@ private void reportPurityError(String msgKeyPrefix, Pair<Tree, String> r) {
         }
     }
 
+    /** Matches warnings about use of formal parameter name. */
```

Comment:
It would be helpful to say what this is related to (I think it's `constructJavaExpressionParseError`).
Consider defining `destructJavaExpressionParseError` there that returns either a tuple of information or null.  That would keep related logic together.

Here is a different approach that I think would be better.  `constructJavaExpressionParseError` could take, as an argument, the message key.  Then it can use more specific key and can use different numbers of arguments rather than putting all information in a string that needs to be parsed later.
Currently, the only two message keys used with JavaExpressionParseException are "flowexpr.parse.index.too.big" and "flowexpr.parse.error", but there could be more.  That would permit them to be more specific and would lead to more readable error messages and better handling in catch blocks.
This pull request uses "expression.parameter.name.invalid", which could be used deeper in the code.

(I'm sorry that this leads to more refactoring, but I think it will be an improvement overall.)

Reply:
I didn't take this approach, but I changed it so there isn't any string parsing.  I agree that it would be nice to issue different parse errors, but perhaps we can do that later.